File: sdk/villow-sdk-0.1.0/package/src/questions.ts
export type JsonObject = Record<string, unknown>;

export interface BaseQuestion {
  id: string;
  type: string;
  label: string;
  required: boolean;
  default_value?: unknown;
  validation: JsonObject;
  decide_for_me: JsonObject;
  options?: string[];
}

interface BaseQuestionInput {
  id: string;
  label: string;
  default?: unknown;
  decideForMe?: JsonObject;
  required?: boolean;
}

export class Question {
  static singleSelect(input: BaseQuestionInput & { options: string[]; default?: string }): BaseQuestion {
    if (input.options.length === 0) {
      throw new Error('singleSelect requires at least one option');
    }
    if (input.default !== undefined && !input.options.includes(input.default)) {
      throw new Error('default must be one of options');
    }
    return question({
      ...input,
      type: 'single_select',
      validation: { options: input.options },
      options: input.options,
    });
  }

  static number(input: BaseQuestionInput & { min?: number; max?: number; default?: number }): BaseQuestion {
    if (input.min !== undefined && input.max !== undefined && input.min > input.max) {
      throw new Error('min cannot be greater than max');
    }
    if (input.default !== undefined) {
      if (input.min !== undefined && input.default < input.min) {
        throw new Error('default cannot be below min');
      }
      if (input.max !== undefined && input.default > input.max) {
        throw new Error('default cannot exceed max');
      }
    }
    const validation: JsonObject = {};
    if (input.min !== undefined) validation.min = input.min;
    if (input.max !== undefined) validation.max = input.max;
    return question({ ...input, type: 'number', validation });
  }

  static boolean(input: BaseQuestionInput & { default?: boolean }): BaseQuestion {
    return question({ ...input, type: 'boolean', validation: {} });
  }

  static text(input: BaseQuestionInput & { default?: string; maxLength?: number }): BaseQuestion {
    const validation: JsonObject = {};
    if (input.maxLength !== undefined) validation.max_length = input.maxLength;
    return question({ ...input, type: 'text', validation });
  }
}

function question(
  input: BaseQuestionInput & { type: string; validation: JsonObject; options?: string[] },
): BaseQuestion {
  if (!input.id) throw new Error('question id is required');
  if (!input.label) throw new Error('question label is required');
  if (!input.decideForMe) throw new Error('decideForMe is required for every clarification question');
  const payload: BaseQuestion = {
    id: input.id,
    type: input.type,
    label: input.label,
    required: input.required ?? true,
    validation: input.validation,
    decide_for_me: input.decideForMe,
  };
  if (input.default !== undefined) payload.default_value = input.default;
  if (input.options) payload.options = input.options;
  return payload;
}

File: sdk/villow-0.1.0/villow-0.1.0/src/villow/questions.py
from __future__ import annotations

from typing import Any


class Question:
    @staticmethod
    def single_select(
        *,
        id: str,
        label: str,
        options: list[str],
        default: str | None = None,
        decide_for_me: dict[str, Any] | None,
        required: bool = True,
    ) -> dict[str, Any]:
        if not options:
            raise ValueError("single_select requires at least one option")
        if default is not None and default not in options:
            raise ValueError("default must be one of options")
        return _question(
            id=id,
            label=label,
            question_type="single_select",
            default=default,
            validation={"options": options},
            options=options,
            decide_for_me=decide_for_me,
            required=required,
        )

    @staticmethod
    def number(
        *,
        id: str,
        label: str,
        min: int | float | None = None,
        max: int | float | None = None,
        default: int | float | None = None,
        decide_for_me: dict[str, Any] | None,
        required: bool = True,
    ) -> dict[str, Any]:
        validation: dict[str, Any] = {}
        if min is not None:
            validation["min"] = min
        if max is not None:
            validation["max"] = max
        if min is not None and max is not None and min > max:
            raise ValueError("min cannot be greater than max")
        if default is not None:
            if min is not None and default < min:
                raise ValueError("default cannot be below min")
            if max is not None and default > max:
                raise ValueError("default cannot exceed max")
        return _question(
            id=id,
            label=label,
            question_type="number",
            default=default,
            validation=validation,
            decide_for_me=decide_for_me,
            required=required,
        )

    @staticmethod
    def boolean(
        *,
        id: str,
        label: str,
        default: bool | None = None,
        decide_for_me: dict[str, Any] | None,
        required: bool = True,
    ) -> dict[str, Any]:
        return _question(
            id=id,
            label=label,
            question_type="boolean",
            default=default,
            validation={},
            decide_for_me=decide_for_me,
            required=required,
        )

    @staticmethod
    def text(
        *,
        id: str,
        label: str,
        default: str | None = None,
        decide_for_me: dict[str, Any] | None,
        required: bool = True,
        max_length: int | None = None,
    ) -> dict[str, Any]:
        validation: dict[str, Any] = {}
        if max_length is not None:
            validation["max_length"] = max_length
        return _question(
            id=id,
            label=label,
            question_type="text",
            default=default,
            validation=validation,
            decide_for_me=decide_for_me,
            required=required,
        )


def _question(
    *,
    id: str,
    label: str,
    question_type: str,
    default: Any,
    validation: dict[str, Any],
    decide_for_me: dict[str, Any] | None,
    required: bool,
    options: list[str] | None = None,
) -> dict[str, Any]:
    if not id:
        raise ValueError("question id is required")
    if not label:
        raise ValueError("question label is required")
    if decide_for_me is None:
        raise ValueError("decide_for_me is required for every clarification question")
    question: dict[str, Any] = {
        "id": id,
        "type": question_type,
        "label": label,
        "required": required,
        "default_value": default,
        "validation": validation,
        "decide_for_me": decide_for_me,
    }
    if options is not None:
        question["options"] = options
    return question
